diff -ur a/ricci/modules/cluster/ClusterStatus.cpp b/ricci/modules/cluster/ClusterStatus.cpp
--- a/ricci/modules/cluster/ClusterStatus.cpp	2011-06-20 12:19:54.000000000 -0400
+++ b/ricci/modules/cluster/ClusterStatus.cpp	2011-06-20 12:23:16.000000000 -0400
@@ -34,7 +34,8 @@
 
 #define INITD_DIR_PATH			"/etc/init.d/"
 #define CHKCONFIG_PATH			"/sbin/chkconfig"
-#define LVMCONF_PATH			"/usr/sbin/lvmconf"
+#define LVMCONF_PATH			"/sbin/lvmconf"
+#define UMOUNT_PATH				"/bin/umount"
 
 #define CMAN_LEAVE_TIMEOUT		"120"	// seconds (string)
 #define CLUMON_SYNC_TIME		8		// seconds
@@ -207,8 +208,27 @@
 	}
 
 	if (stat.get_attr("cluster_version") == "6") {
+		vector<String> args;
+		String out, err;
+		int status;
+
 		run_initd("rgmanager", false, true);
 		run_initd("gfs2", false, false);
+
+		/*
+		** Manually unmount any gfs2 filesystems not umounted by
+		** either rgmanager of the gfs2 init script.
+		*/
+		args.push_back("-a");
+		args.push_back("-t");
+		args.push_back("gfs2");
+
+		if (utils::execute(UMOUNT_PATH, args, out, err, status, false))
+			throw command_not_found_error_msg(UMOUNT_PATH);
+
+		if (status != 0)
+			throw String("Error unmounting gfs2 filesystems: " + err);
+
 		run_initd("clvmd", false, false);
 		run_initd("cman", false, true);
 
Only in b/ricci/modules/cluster: ClusterStatus.cpp.orig
Only in b/ricci/modules/cluster: ClusterStatus.cpp.rej
