commit 4289eddb478b84f83a823181e76d0e38598cd1c4
Author: Ryan McCabe <rmccabe@redhat.com>
Date:   Thu Apr 21 16:06:01 2011 -0400

    Add get_cluster_schema call

diff --git a/ricci/modules/cluster/ClusterModule.cpp b/ricci/modules/cluster/ClusterModule.cpp
index 3e23101..0267d98 100644
--- a/ricci/modules/cluster/ClusterModule.cpp
+++ b/ricci/modules/cluster/ClusterModule.cpp
@@ -35,7 +35,7 @@
 
 using namespace std;
 
-
+static VarMap get_cluster_schema(const VarMap& args);
 static VarMap get_cluster_conf(const VarMap& args);
 static VarMap set_cluster_conf(const VarMap& args);
 static VarMap set_cluster_version(const VarMap& args);
@@ -73,6 +73,7 @@ build_fcn_map()
 	api_1_0["get_cluster.conf"]		= get_cluster_conf;
 	api_1_0["set_cluster.conf"]		= set_cluster_conf;
 	api_1_0["set_cluster_version"]	= set_cluster_version;
+	api_1_0["get_cluster_schema"]	= get_cluster_schema;
 
 	api_1_0["status"]				= cluster_status;
 
@@ -110,6 +111,17 @@ cluster_status(const VarMap& args)
 }
 
 VarMap
+get_cluster_schema(const VarMap& args)
+{
+
+	Variable var("cluster_schema", readXML(CLUSTER_SCHEMA_PATH));
+
+	VarMap ret;
+	ret.insert(pair<String, Variable>(var.name(), var));
+	return ret;
+}
+
+VarMap
 get_cluster_conf(const VarMap& args)
 {
 	Variable var("cluster.conf", ClusterConf::get());
diff --git a/ricci/modules/cluster/ClusterModule.h b/ricci/modules/cluster/ClusterModule.h
index 5541a90..0d69a0b 100644
--- a/ricci/modules/cluster/ClusterModule.h
+++ b/ricci/modules/cluster/ClusterModule.h
@@ -23,6 +23,8 @@
 #ifndef __CONGA_MODCLUSTER_CLUSTERMODULE_H
 #define __CONGA_MODCLUSTER_CLUSTERMODULE_H
 
+#define CLUSTER_SCHEMA_PATH     "/usr/share/cluster/cluster.rng"
+
 #include "Module.h"
 
 class ClusterModule : public Module
diff --git a/ricci/test_suite/cluster/get_cluster_schema.xml b/ricci/test_suite/cluster/get_cluster_schema.xml
new file mode 100644
index 0000000..8712890
--- /dev/null
+++ b/ricci/test_suite/cluster/get_cluster_schema.xml
@@ -0,0 +1,13 @@
+<?xml version="1.0" ?>
+<ricci version="1.0" function="process_batch" async="false">
+<batch>
+
+<module name="cluster">
+<request sequence="1254" API_version="1.0">
+<function_call name="get_cluster_schema"/>
+</request>
+</module>
+
+</batch>
+</ricci>
+
