diff -up clustermon-0.16.2/ricci/modules/cluster/Clusvcadm.cpp.orig clustermon-0.16.2/ricci/modules/cluster/Clusvcadm.cpp
--- clustermon-0.16.2/ricci/modules/cluster/Clusvcadm.cpp.orig	2011-08-23 14:50:50.381749373 -0400
+++ clustermon-0.16.2/ricci/modules/cluster/Clusvcadm.cpp	2011-08-23 16:43:37.732934030 -0400
@@ -148,6 +148,9 @@ void Clusvcadm::start(const String& serv
 			args.push_back("-m");
 			args.push_back(nodename);
 		}
+		else {
+			args.push_back("-F");
+		}	
 
 		if (utils::execute(CLUSVCADM_TOOL_PATH, args, out, err, status, false))
 			throw command_not_found_error_msg(CLUSVCADM_TOOL_PATH);
