diff -ur a/ricci/common/XML.cpp b/ricci/common/XML.cpp
--- a/ricci/common/XML.cpp
+++ b/ricci/common/XML.cpp
@@ -208,29 +208,38 @@
 			// kids
 			_parseXML(me, curr_node->children);
 			parent.add_child(me);
-		} else if (curr_node->type == XML_COMMENT_NODE) {
-			xmlChar *comment_content = xmlNodeGetContent(curr_node);
+		} else if ((curr_node->type == XML_COMMENT_NODE || curr_node->type == XML_TEXT_NODE)
+					&& !xmlIsBlankNode(curr_node))
+		{
+			xmlChar *node_content = xmlNodeGetContent(curr_node);
 
-			if (comment_content) {
-				String text((const char *) comment_content);
-				xmlFree((void *) comment_content);
+			if (node_content) {
+				char *tp = (char *) node_content;
+				size_t content_len = strlen(tp);
+				char *text_start;
+				char *text_end;
 
-				XMLObject me("COMMENT");
-				me.set_node_type(XML_COMMENT_NODE);
+				if (content_len > 0)
+					text_end = &tp[content_len - 1];
+				else
+					text_end = tp;
+
+				while (isspace(*tp))
+					tp++;
+				text_start = tp;
+
+				while (text_end > text_start && isspace(*text_end))
+					text_end--;
+				*(text_end + 1) = '\0';
+
+				String text((const char *) text_start);
+				xmlFree((void *) node_content);
+
+				XMLObject me("DUMMY");
+				me.set_node_type(curr_node->type);
 				me.set_text(text);
 				parent.add_child(me);
 			}
-		} else if (curr_node->type == XML_TEXT_NODE && !xmlIsBlankNode(curr_node)) {
-			xmlChar *content = xmlNodeGetContent(curr_node);
-			if (content) {
-				String node_content((const char *) content);
-				xmlFree(content);
-
-				XMLObject me("TEXT");
-				me.set_node_type(XML_TEXT_NODE);
-				me.set_text(node_content);
-				parent.add_child(me);
-			}
 		}
 	}
 }
