--- a/ricci/common/File.cpp.orig	2012-11-19 20:40:31.635710508 +0100
+++ b/ricci/common/File.cpp	2012-11-19 21:37:23.082560218 +0100
@@ -138,18 +138,28 @@
 {
 	MutexLocker l(*_mutex);
 
+	char *buff = NULL;
 	long len = size();
-	char buff[len];
+
 	try {
+		buff = new char[len];
+
 		((fstream *) _pimpl->fs)->seekg(0, ios::beg);
-		check_failed();
 		((fstream *) _pimpl->fs)->read(buff, len);
 		check_failed();
+		if (len != ((fstream *) _pimpl->fs)->gcount())
+			throw String("Read size mismatch: ") + _path;
 		String ret(buff, len);
+
 		::shred(buff, len);
+		delete[] buff;
+
 		return ret;
 	} catch ( ... ) {
-		::shred(buff, len);
+		if (buff) {
+			::shred(buff, len);
+			delete[] buff;
+		}
 		throw;
 	}
 }
